modules/gapi/src/backends/oak/goakbackend.cpp

```diff
@@ -0,0 +1,704 @@
+                                     std::vector<ExtractTypeHelper::InputPtr>& oak_ins) {
+    if (arg.kind != cv::detail::ArgKind::GOBJREF) {
+        GAPI_Assert(   arg.kind != cv::detail::ArgKind::GMAT
+                    && arg.kind != cv::detail::ArgKind::GSCALAR
+                    && arg.kind != cv::detail::ArgKind::GARRAY
+                    && arg.kind != cv::detail::ArgKind::GOPAQUE
+                    && arg.kind != cv::detail::ArgKind::GFRAME);
+        // All other cases - pass as-is, with no transformations to
+        // GArg contents.
+        return const_cast<cv::GArg&>(arg);
+    }
+    const cv::gimpl::RcDesc &ref = arg.get<cv::gimpl::RcDesc>();
+    switch (ref.shape) {
+    case GShape::GFRAME:
+        oak_ins.push_back(nullptr);
+        return GArg(std::reference_wrapper<ExtractTypeHelper::InputPtr>(oak_ins.back()));
+        break;
+    default:
+        util::throw_error(std::logic_error("Unsupported GShape type in OAK backend"));
+        break;
+    }
+}
+
+void cv::gimpl::GOAKExecutable::packOutArg(const RcDesc &rc,
+                                           std::vector<ExtractTypeHelper::OutputPtr>& oak_outs) {
+    switch (rc.shape) {
+    case GShape::GFRAME:
+        oak_outs.push_back(nullptr);
+        break;
+    case GShape::GARRAY:
+        oak_outs.push_back(nullptr);
+        break;
+    default:
+        util::throw_error(std::logic_error("Unsupported GShape type in OAK backend"));
+        break;
+    }
+}
+
+cv::gimpl::GOAKExecutable::GOAKExecutable(const ade::Graph& g,
+                                          const cv::GCompileArgs &args,
+                                          const std::vector<ade::NodeHandle>& nodes,
+                                          const std::vector<cv::gimpl::Data>& ins_data,
+                                          const std::vector<cv::gimpl::Data>& outs_data)
+    : m_g(g), m_gm(m_g), m_args(args),
+      m_device(nullptr), m_pipeline(new dai::Pipeline)
+    {
+        // FIXME: currently OAK backend only works with camera as input,
+        //        so it must be a single object
+        GAPI_Assert(ins_data.size() == 1);
+
+        // Check that there is only one OAK island in graph since there
+        // can only be one instance of dai::Pipeline in the application
+        auto isl_graph = m_gm.metadata().get<IslandModel>().model;
+        GIslandModel::Graph gim(*isl_graph);
+        size_t oak_islands = 0;
+
+        for (const auto& nh : gim.nodes())
+        {
+            if (gim.metadata(nh).get<NodeKind>().k == NodeKind::ISLAND)
+            {
+                const auto isl = gim.metadata(nh).get<FusedIsland>().object;
+                if (isl->backend() == cv::gapi::oak::backend())
+                {
+                    ++oak_islands;
+                }
+            }
+        }
+
+        if (oak_islands != 1) {
+            GAPI_Assert(false && "There can only be a single OAK island in graph");
```

Comment:
Better to explicitly throw an exception here, and you can throw earlier, just at the moment when the second oak island is found

Reply:
Done